apps/semantic-cache/src/index.ts

```diff
@@ -0,0 +1,363 @@
+    this.isComplete = false;
+  }
+
+  async readToEnd() {
+    try {
+      while (true) {
+        const { done, value } = await this.reader.read();
+        if (done) {
+          this.isDone = true;
+          break;
+        }
+        this.data += new TextDecoder().decode(value);
+      }
+    } catch (error) {
+      console.error("Stream error:", error);
+      this.isDone = false;
+    } finally {
+      this.reader.releaseLock();
+    }
+    return this.isDone;
+  }
+
+  checkComplete() {
+    if (this.data.includes("[DONE]")) {
+      this.isComplete = true;
+    }
+  }
+
+  getReader() {
+    return this.reader;
+  }
+
+  getData() {
+    return this.data;
+  }
+}
+
+async function handleCacheOrDiscard(
+  c: Context,
+  cache: UnkeyCache<{ response: Response }>,
+  stream: ManagedStream,
+  event: AnalyticsEvent,
+  vector?: number[],
+) {
+  await stream.readToEnd();
+
+  // Check if the data is complete and should be cached
+  if (stream.isDone) {
+    const id = nanoid();
+    const rawData = stream.getData();
+    let contentStr = "";
+    for (const token of rawData.split("\n")) {
+      contentStr += extractWord(token);
+    }
+    const time = Date.now();
+    await cache.response.set(id, { id, content: contentStr });
+    const writeTime = Date.now();
+    console.info(`Cached with ID: ${id}, time: ${writeTime - time}ms`);
+    if (vector) {
+      await c.env.VECTORIZE_INDEX.insert([{ id, values: vector }]);
+    }
+
+    const analytics = new Analytics({ tinybirdToken: c.env.TINYBIRD_TOKEN });
+    const finalEvent = {
+      ...event,
+      cache: true,
+      requestId: id,
+      timing: writeTime - time,
+      tokens: rawData.split("\n").length,
+      response: contentStr,
+    };
+    analytics
+      .ingestLogs(finalEvent)
+      .then(() => {
+        console.info("Logs persisted in Tinybird");
+      })
+      .catch((err) => {
+        console.error("Error persisting logs in Tinybird:", err);
+      });
+    console.info("Data cached in KV with ID:", id);
+  } else {
+    console.info("Data discarded, did not end properly.");
+  }
+}
+
+async function handleStreamingRequest(
+  c: Context,
+  request: OpenAI.Chat.Completions.ChatCompletionCreateParamsStreaming & {
+    noCache?: boolean;
+  },
+  openai: OpenAI,
+  cache: UnkeyCache<{ response: Response }>,
+) {
+  c.header("Connection", "keep-alive");
+  c.header("Cache-Control", "no-cache, must-revalidate");
+
+  const startTime = Date.now();
+  const messages = parseMessagesToString(request.messages);
+  console.info("Messages:", messages);
+  const vector = await getEmbeddings(c, messages);
+  const embeddingsTime = Date.now();
+  const query = await c.env.VECTORIZE_INDEX.query(vector, { topK: 1 });
+  const queryTime = Date.now();
+
+  const event = {
+    timestamp: new Date().toISOString(),
+    model: request.model,
+    stream: request.stream,
+    query: messages,
+    vector: [0],
+  };
+
+  console.info("Query results:", query);
```

Comment:
please use a logger similar to how we do it in the API, we need everything to be transported to axiom

Reply:
OK to do this in a separate PR? I'm working on this but would like to avoid this one becoming massive